### PR TITLE
L3 231/add monitoring

### PR DESCRIPTION
**Pull request by msterle**
PR 1/2 for L3-231

This PR adds monitoring to the king cluster, production cluster addressed in a follow-up PR

- Added kube-prometheus-stack chart for monitoring.
- Added ServiceMonitors for:
  - sqnc-node (for alice, bob and charlie)
  - sqnc-identity-service postgresql (for alice, bob and charlie)
  - sqnc-matchmaker-api postgresql (for alice, bob and charlie)
  - flux components
  - self-monitoring of monitoring components
- Add grafana dashboards for:
  - alertmanager
  - flux cluster stats (unsure if working, pulled from flux documentation)
  - flux control plane (pulled from flux documentation)
  - Flux2 (from grafana.com)
  - nginx (from grafana.com)
  - polkadot (from grafana.com)
  - postgresql (from grafana.com)

Notes:
- Currently, the flux cluster stats dashboard is broken due to [a bug in the `kube-state-metrics` chart](https://github.com/kubernetes/kube-state-metrics/issues/1973). This bug has been resolved, but the fix has not been pulled into `kube-prometheus-stack` yet

### CODE REVIEW

**Review by dblane-digicatapult — CHANGES_REQUESTED** [review]
This looks good, we should ensure that alice, bob, charlie and the nginx/infra sync depend on the monitoring sync kustomization being completed first as this introduces the CRDs that will subsequently be used in those kustomizations/helm charts.

File: clusters/kind-cluster/base/app-sync.yaml [hunk 1]
```diff
@@ -45,11 +51,27 @@ apiVersion: kustomize.toolkit.fluxcd.io/v1
 kind: Kustomization
 metadata:
   name: nginx-sync
-  namespace: kube-system
+  namespace: sqnc
```
Comment (dblane-digicatapult) [inline clusters/kind-cluster/base/app-sync.yaml:54]: Why has this NS changed?

File: clusters/kind-cluster/base/namespaces.yaml [hunk 2]
```diff
@@ -24,8 +24,24 @@ metadata:
 ---
 apiVersion: v1
 kind: Namespace
+metadata:
+  labels:
+    app.kubernetes.io/instance: sqnc
+    app.kubernetes.io/version: latest
+  name: sqnc
```
Comment (dblane-digicatapult) [inline clusters/kind-cluster/base/namespaces.yaml:31]: What is the purpose of this NS?

File: clusters/kind-cluster/monitoring/kube-state-metrics-config.yaml
Comment (dblane-digicatapult) [inline clusters/kind-cluster/monitoring/kube-state-metrics-config.yaml:1]: Could you add a comment that gives us the source location for this file and how to configure it?

File: clusters/kind-cluster/monitoring/release.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
+apiVersion: helm.toolkit.fluxcd.io/v2beta2
+kind: HelmRelease
+metadata:
+  name: kube-prometheus-stack
+spec:
+  interval: 1h
```
Comment (dblane-digicatapult) [inline clusters/kind-cluster/monitoring/release.yaml:6]: This interval is too long. 10m is fine.

File: clusters/kind-cluster/monitoring/release.yaml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  chart:
+    spec:
+      version: "57.x"
+      chart: kube-prometheus-stack
+      sourceRef:
+        kind: HelmRepository
+        name: prometheus-community
+      interval: 1h
```
Comment (dblane-digicatapult) [inline clusters/kind-cluster/monitoring/release.yaml:14]:
```suggestion
```

File: clusters/kind-cluster/monitoring/release.yaml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    crds: Create
+  upgrade:
+    crds: CreateReplace
+  driftDetection:
+    mode: enabled
+    ignore:
+      # Ignore "validated" annotation which is not inserted during install
+      - paths: [ "/metadata/annotations/prometheus-operator-validated" ]
+        target:
+          kind: PrometheusRule
```
Comment (dblane-digicatapult) [inline clusters/kind-cluster/monitoring/release.yaml:25]: Under what circumstances does this occur aside installation?

File: clusters/kind-cluster/monitoring/source.yaml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+apiVersion: source.toolkit.fluxcd.io/v1beta2
+kind: HelmRepository
+metadata:
+  name: prometheus-community
+spec:
+  interval: 12h
```
Comment (dblane-digicatapult) [inline clusters/kind-cluster/monitoring/source.yaml:6]: This interval can be decreased

File: clusters/kind-cluster/nginx/release.yaml [hunk 7]
```diff
@@ -3,7 +3,7 @@ apiVersion: helm.toolkit.fluxcd.io/v2beta2
 kind: HelmRelease
 metadata:
   name: nginx
-  namespace: kube-system
+  namespace: sqnc
```
Comment (dblane-digicatapult) [inline clusters/kind-cluster/nginx/release.yaml:6]: Is there any reason why we are moving from this NS?

File: scripts/flux-suspend-all.sh [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+flux_suspend () {
+  case $# in
+    3) flux suspend $1 $3 -n $2;;
+    4) flux suspend $1 $2 $4 -n $3;;
+    *) printf "Expected 3 or 4 parameters, got %s\n" "$#" >&2; exit 1;;
+  esac
+}
+
+export -f flux_suspend
```
Comment (dblane-digicatapult) [inline scripts/flux-suspend-all.sh:11]: Same comment as flux-resume-all.sh

**Review by dblane-digicatapult — CHANGES_REQUESTED** [review]

File: scripts/flux-resume-all.sh [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  case $# in
+    3) flux resume $1 $3 -n $2;;
+    4) flux resume $1 $2 $4 -n $3;;
+    *) printf "Expected 3 or 4 parameters, got %s\n" "$#" >&2; exit 1;;
+  esac
+}
+
+export -f flux_resume
+
+flux get source git -A --no-header --cluster kind-sqnc-flux-infra | awk -F '\ *\t' -v OFS='\t' '$3 ~ /^[:space]*$/ { $3 = "X" } 1' | awk '$4=="True" { print $1,$2 }' | xargs -I {} bash -c 'flux_resume source git {}'
```
[the remark this answers is not in the document]
Comment (msterle) [reply scripts/flux-resume-all.sh:13]: I've actually stopped using the approach of locally updating the cluster, so I'm not sure if it's worth refining these scripts, or removing them entirely as of little value

File: clusters/kind-cluster/monitoring/release.yaml (hunk 5 above)
Comment (msterle) [reply clusters/kind-cluster/monitoring/release.yaml:25]: This should only occur during installation as far as I can tell: https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/admission/admission.go#L211-L212

**Review by dblane-digicatapult — APPROVED** [review]

**Review by dblane-digicatapult — APPROVED** [review]

**Pull request merged**